app/services/samples/metadata/file_import_service.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+require 'roo'
+
+module Samples
+  module Metadata
+    # Service used to import sample metadata via a file
+    class FileImportService < BaseService
+      SampleMetadataFileImportError = Class.new(StandardError)
+
+      def initialize(project, user = nil, params = {})
+        super(user, params)
+        @project = project
+        @file = params[:file]
+        @sample_id_column = params[:sample_id_column]
+        @ignore_empty_values = params[:ignore_empty_values] || false
+        @spreadsheet = nil
+        @headers = nil
+      end
+
+      def execute
+        authorize! @project, to: :update_sample?
+
+        validate_sample_id_column
+
+        validate_file
+
+        perform_file_import
+      rescue Samples::Metadata::FileImportService::SampleMetadataFileImportError => e
+        @project.errors.add(:base, e.message)
+        {}
+      end
+
+      private
+
+      def validate_sample_id_column
+        return unless @sample_id_column.nil?
+
+        raise SampleMetadataFileImportError,
+              I18n.t('services.samples.metadata.import_file.empty_sample_id_column')
+      end
+
+      def validate_file # rubocop:disable Metrics/MethodLength, Metrics/AbcSize, Metrics/CyclomaticComplexity
+        if @file.nil?
+          raise SampleMetadataFileImportError,
+                I18n.t('services.samples.metadata.import_file.empty_file')
+        end
+
+        file_extension = File.extname(@file).downcase
```

Comment:
A little refactoring to possibly remove some of the rubocop disables on this method. We can move the file extension validation into it's own method, the validation of the file contents (opening and checking for the columns) can be moved into it's own method as well

Reply:
0455419f8ab52f8a4ce8956eda4a29c21195e97b